src/coreclr/jit/emitxarch.cpp

```diff
@@ -596,26 +611,48 @@ bool emitter::AreUpper32BitsSignExtended(regNumber reg)
 
     instrDesc* id = emitLastIns;
 
-    if (id->idReg1() != reg)
-    {
-        return false;
-    }
+    bool result = false;
 
-    // movsx always sign extends to 8 bytes. W-bit is set.
-    if (id->idIns() == INS_movsx)
-    {
-        return true;
-    }
+    emitPeepholeIterateLastInstrs([&](instrDesc* id) {
+        if (emitIsInstrWritingToReg(id, reg))
+        {
+            switch (id->idIns())
+            {
+                // Conservative.
+                case INS_call:
+                    return PEEPHOLE_ABORT;
 
-    // movsxd is always an 8 byte operation. W-bit is set.
-    if (id->idIns() == INS_movsxd)
-    {
-        return true;
-    }
+                case INS_movsx:
+#ifdef TARGET_64BIT
+                case INS_movsxd:
+#endif // TARGET_64BIT
+                    if ((size == EA_1BYTE) || (size == EA_2BYTE))
+                    {
+                        result = (id->idOpSize() <= size);
+                    }
+#ifdef TARGET_64BIT
+                    // movsx/movsxd always sign extends to 8 bytes. W-bit is set.
```

Comment:
Why does the `W-bit` matter here?

Reply:
W-bit on the encoding is what makes `movsx` and `movsxd` sign-extend to 8 bytes. We always default to using the W-bit.

Reply:
This function cares about instructions and sizes, but not encoding, though.

Reply:
The comment `W-bit is set` itself is just meant to inform the reason why this optimization is safe to do. It was already there before.